So I'm running some tests in a fresh 9front QEMU image.

I ran into this during `go build -o lg .`:

<img width="2804" height="1322" alt="image" src="https://github.com/user-attachments/assets/e4413897-e5a6-478f-9c97-549ed3e521e1" />

I didn't run into that when I build on agent9.
Not sure if something changed or if it's something to do with this particular image.
Just thought I'd share the issue.

Full terminal log shown below:

```
cpu% echo ipv6 >/net/cs
cpu% webfs
cpu% 
cpu% hget https://raw.githubusercontent.com/Alino/agent9/main/pac9/pac9 >/rc/bin/pac9
cpu% chmod +x /rc/bin/pac9
cpu% mkdir -p /sys/lib/pac9
cpu% hget https://raw.githubusercontent.com/Alino/agent9/main/pac9/registry >/sys/lib/pac9/registry
cpu% 
cpu% pac9 install go
pac9: fetching https://github.com/Alino/agent9/releases/download/go/go-amd64.tar.gz
pac9: installed go
cpu% go
Go is a tool for managing Go source code.

Usage:

	go <command> [arguments]

The commands are:

	bug         start a bug report
	build       compile packages and dependencies
	clean       remove object files and cached files
	doc         show documentation for package or symbol
	env         print Go environment information
	fix         apply fixes suggested by static checkers
	fmt         gofmt (reformat) package sources
	generate    generate Go files by processing source
	get         add dependencies to current module and install them
	install     compile and install packages and dependencies
	list        list packages or modules
	mod         module maintenance
	work        workspace maintenance
	run         compile and run Go program
	telemetry   manage telemetry data and settings
	test        test packages
	tool        run specified go tool
	version     print Go version
	vet         report likely mistakes in packages

Use "go help <command>" for more information about a command.

Additional help topics:

	buildconstraint build constraints
	buildjson       build -json encoding
	buildmode       build modes
	c               calling between Go and C
	cache           build and test caching
	environment     environment variables
	filetype        file types
	goauth          GOAUTH environment variable
	go.mod          the go.mod file
	gopath          GOPATH environment variable
	goproxy         module proxy protocol
	importpath      import path syntax
	modules         modules, module versions, and more
	module-auth     module authentication using go.sum
	packages        package lists and patterns
	private         configuration for downloading non-public code
	testflag        testing flags
	testfunc        testing functions
	vcs             controlling version control with GOVCS

Use "go help <topic>" for more information about that topic.

cpu% cd
cpu% mkdir src
cpu% pwd
/usr/glenda
cpu% cd src
cpu% pwd
/usr/glenda/src
cpu% lc
cpu% git/clone https://github.com/nooga/let-go.git
fetching... /
indexing 9268 objects: 100%
checking out repository...
cpu% cd let-go
cpu% lc
.agents/				.goreleaser.yml			README.md				lg_ansi.go				pkg/
.claude/				.pre-commit-config.yaml	benchmark/				lg_ansi_plan9.go		scripts/
.git/					.vscode/				cmd/					lg_profile.go			tap_migrations.json
.gitattributes			CONTRIBUTORS			deps.edn				lg_profile_default.go	test/
.github/				Casks/					docs/					lg_repl.go				wasm/
.gitignore				Dockerfile				examples/				lg_repl_stub.go			wasm.go
.gitmodules				HomebrewFormula/		go.mod					meta/
.golangci.yml			LICENSE					go.sum					mise.toml
.gomodcache/			Makefile				lg.go					openspec/
cpu% git/branch -a
heads/main
remotes/origin/agent/fix-plan9-key-pending
remotes/origin/agent/plan9-terminal-input
remotes/origin/ci/dedicated-determinism-step
remotes/origin/codex/key-tokenizer-fsm
remotes/origin/delint-golangci-v2
remotes/origin/docs-status-skill
remotes/origin/docs/docs-status-tool
remotes/origin/feat/ir-substrate-passes
remotes/origin/feat/js-url-param
remotes/origin/feat/set-rand-seed
remotes/origin/fix-http-header-raw-strings
remotes/origin/fix-var-doc-metadata
remotes/origin/fix/426-update-keys-vals-re-find
remotes/origin/fix/ns-loader-require-semantics
remotes/origin/fix/pod-shutdown-without-op
remotes/origin/fix/source-bootstrap-lgbgen
remotes/origin/homebrew-tap-let-go
remotes/origin/main
remotes/origin/migrate-homebrew-cask
remotes/origin/perf-data
remotes/origin/perf-timeline-page
remotes/origin/perf/lower-go-registry-exact-resolve
remotes/origin/perf/pr-fast-tune
remotes/origin/pr/wasm-host-eval
remotes/origin/release-brews-arm-linux
remotes/origin/release-workflow-tag-branch-fix
remotes/origin/release/v1.9.0
remotes/origin/runtime-only-build
remotes/origin/spike/libyamlstar-cshared
remotes/origin/sweep/feature/http-request-function
remotes/origin/sweep/reader-unit-tests
remotes/origin/tco
remotes/origin/wasip1-build
cpu% 
cpu% git/branch agent/plan9-terminal-input
refs/heads/agent/plan9-terminal-input: ccff61420c1415507c10747d13ca729f5fbacda1
cpu% go build -o lg .
go: downloading github.com/hashicorp/go-uuid v1.0.3
go: downloading github.com/zeebo/bencode v1.0.0
go: downloading github.com/zeebo/xxh3 v1.1.0
pkg/rt/pods.go:21:2: unable to resolve git version: failed to execute git version: exec: "git": executable file not found in $path
pkg/rt/interop_xxh3.go:6:2: unable to resolve git version: failed to execute git version: exec: "git": executable file not found in $path
pkg/nrepl/server.go:17:2: unable to resolve git version: failed to execute git version: exec: "git": executable file not found in $path
```

--- a/pkg/rt/keysource_queued.go
+++ b/pkg/rt/keysource_queued.go
@@ -1,0 +1,132 @@
+/*
+ * Copyright (c) 2026 Matt Parrett
+ * SPDX-License-Identifier: MIT
+ */
+
+package rt
+
+import (
+	"io"
+	"sync"
+)
+
+// queuedKeySource is a KeySource for hosts without a non-blocking input peek: a
+// background goroutine blocks on an io.Reader and appends to a shared queue,
+// ReadKey tokenizes one key per call off it (via scanKey), and KeyPending peeks
+// it. This is the shape Plan 9 needs — it has no poll(2)/FIONREAD, so a
+// non-blocking key-pending? can only be answered from a buffer a concurrent
+// reader fills, where native (term.go) reads stdin synchronously inside ReadKey.
+// It is platform-neutral, so it is also available to embedders whose input is a
+// plain blocking reader (a pipe, a socket) — see NewQueuedKeySource.
+//
+// State lives on the instance, not a package global like native's keyBuf: the
+// goroutine consumes the reader, so it must start only when THIS source is the
+// one being read. If an embedder rebinds *keys* via api.WithKeySource, this
+// source's ReadKey/KeyPending are never called and the reader never starts — no
+// bytes stolen from the source that replaced it.
+type queuedKeySource struct {
+	mu      sync.Mutex
+	cond    *sync.Cond
+	r       io.Reader
+	buf     []byte // read but not yet tokenized out
+	eof     bool
+	started bool
+}
+
+// NewQueuedKeySource returns a KeySource that reads keys from r via a background
+// goroutine and a buffered queue, so KeyPending is non-blocking on platforms
+// that lack a poll/FIONREAD peek. The reader goroutine starts lazily on the
+// first ReadKey/KeyPending, so binding this at *keys* without ever consulting it
+// (e.g. when api.WithKeySource overrides it) reads nothing.
+func NewQueuedKeySource(r io.Reader) KeySource {
+	s := &queuedKeySource{r: r}
+	s.cond = sync.NewCond(&s.mu)
+	return s
+}
+
+// queuedReadChunkSize matches native's readChunkSize: a blocking Read returns as
+// soon as any data is available, so a larger buffer adds no latency to a single
+// keystroke — it just collects held-key bursts in fewer reads.
+const queuedReadChunkSize = 256
+
+// start launches the background reader once. Called with s.mu held.
+func (s *queuedKeySource) start() {
+	if s.started {
+		return
+	}
+	s.started = true
+	go s.readLoop()
+}
+
+// readLoop blocks on r.Read WITHOUT holding the lock, then locks only to append
+// the chunk and wake a waiter. Runs until EOF or a read error, which it records
+// as eof so a parked ReadKey unblocks with the nil contract.
+func (s *queuedKeySource) readLoop() {
+	chunk := make([]byte, queuedReadChunkSize)
+	for {
+		n, err := s.r.Read(chunk)
+		s.mu.Lock()
+		if n > 0 {
+			s.buf = append(s.buf, chunk[:n]...)
+		}
+		if err != nil || n == 0 {
+			s.eof = true
+			s.cond.Broadcast()
+			s.mu.Unlock()
+			return
+		}
+		s.cond.Broadcast()
+		s.mu.Unlock()
+	}
+}
+
+// ReadKey blocks until a whole key token is available (or EOF), tokenizing one
+// key per call off the shared buffer via scanKey. "" signals end-of-input (the
+// read-key nil contract). Blocking here matches native (which blocks in poll);
+// callers stay responsive by gating on KeyPending first.
+func (s *queuedKeySource) ReadKey() (string, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.start()
+	for len(s.buf) == 0 && !s.eof {
+		s.cond.Wait()
+	}
+	if len(s.buf) == 0 {
+		return "", nil // EOF
+	}
+	for {
+		status, n := scanKey(s.buf)
+		if status == keyNeedMore && n > 1 && !s.eof {
+			// A multi-byte token (CSI/SS3/UTF-8) split across reader chunks —
+			// short reads are normal over a network transport like 9P/drawterm.
+			// Wait for one more append, then re-scan to stitch it, rather than
+			// emitting a broken partial that would desync every following key.
+			// Bounded: each wait adds bytes (progress) or hits eof (then emit
+			// best-effort below). A lone ESC (scanKey → keyNeedMore, 1) is
+			// deliberately NOT waited on — it is the Escape key, emitted now
+			// rather than parked until the next keypress.
+			before := len(s.buf)
+			for len(s.buf) == before && !s.eof {
+				s.cond.Wait()
+			}
+			continue
+		}
+		tok := string(s.buf[:n])
+		s.buf = s.buf[n:]
+		if len(s.buf) == 0 {
+			s.buf = nil // release the backing array once drained
+		}
+		return tok, nil
+	}
+}
+
+// KeyPending reports whether a key is buffered and ready, without consuming it.
+// Non-blocking and eof-blind (false once the queue drains at end-of-input),
+// mirroring native's FIONREAD-based rawPending so a per-tick input poll doesn't
+// busy-drain end-of-input forever.
+func (s *queuedKeySource) KeyPending() bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.start()
+	return len(s.buf) > 0
+}

--- a/pkg/rt/keysource_queued.go
+++ b/pkg/rt/keysource_queued.go
@@ -8,6 +8,7 @@ package rt
 import (
 	"io"
 	"sync"
+	"time"
 )
 
 // queuedKeySource is a KeySource for hosts without a non-blocking input peek: a
@@ -26,12 +27,24 @@ import (
 // bytes stolen from the source that replaced it.
 type queuedKeySource struct {
 	mu      sync.Mutex
-	cond    *sync.Cond
 	r       io.Reader
-	buf     []byte // read but not yet tokenized out
-	eof     bool
-	started bool
+	buf     []byte        // read but not yet tokenized out
+	err     error         // sticky reader error; nil for a clean EOF
+	done    bool          // reader finished (EOF or error) — no more bytes coming
+	started bool          // reader goroutine launched
+	notify  chan struct{} // buffered(1) wakeup; reader signals on append/done
+	grace   time.Duration // inter-byte wait for stitching a split escape sequence
 }
+
+// escGrace bounds how long ReadKey waits for the rest of an incomplete escape
+// (or UTF-8) sequence before emitting what it has. It's the classic terminal
+// "ESC delay": long enough to stitch a sequence split across reads on a slow
+// transport (drawterm/9P), short enough that a bare Escape key still registers
+// promptly. Only an incomplete front token pays it — a complete key returns with
+// zero added latency. Without it, distinguishing a bare ESC from the start of a
+// split arrow is impossible on a host with no FIONREAD peek (a lone ESC would
+// either be emitted too eagerly, breaking split arrows, or waited on forever).
+const escGrace = 40 * time.Millisecond
 
 // NewQueuedKeySource returns a KeySource that reads keys from r via a background
 // goroutine and a buffered queue, so KeyPending is non-blocking on platforms
@@ -39,9 +52,7 @@ type queuedKeySource struct {
 // first ReadKey/KeyPending, so binding this at *keys* without ever consulting it
 // (e.g. when api.WithKeySource overrides it) reads nothing.
 func NewQueuedKeySource(r io.Reader) KeySource {
-	s := &queuedKeySource{r: r}
-	s.cond = sync.NewCond(&s.mu)
-	return s
+	return &queuedKeySource{r: r, notify: make(chan struct{}, 1), grace: escGrace}
 }
 
 // queuedReadChunkSize matches native's readChunkSize: a blocking Read returns as
@@ -58,9 +69,21 @@ func (s *queuedKeySource) start() {
 	go s.readLoop()
 }
 
+// signal wakes a waiting ReadKey without blocking; the buffered channel
+// coalesces bursts, and ReadKey re-checks the buffer under the lock after each
+// wake, so a dropped (already-pending) signal is never a missed update.
+func (s *queuedKeySource) signal() {
+	select {
+	case s.notify <- struct{}{}:
+	default:
+	}
+}
+
 // readLoop blocks on r.Read WITHOUT holding the lock, then locks only to append
-// the chunk and wake a waiter. Runs until EOF or a read error, which it records
-// as eof so a parked ReadKey unblocks with the nil contract.
+// the chunk and wake a waiter. On a read error it retains the error (draining
+// any bytes that came with it) and stops; io.EOF is a clean end, recorded as a
+// nil error so ReadKey returns the "" contract. A (0, nil) read is not EOF (the
+// io.Reader contract) — it loops and reads again.
 func (s *queuedKeySource) readLoop() {
 	chunk := make([]byte, queuedReadChunkSize)
 	for {
@@ -69,55 +92,95 @@ func (s *queuedKeySource) readLoop() {
 		if n > 0 {
 			s.buf = append(s.buf, chunk[:n]...)
 		}
-		if err != nil || n == 0 {
-			s.eof = true
-			s.cond.Broadcast()
+		if err != nil {
+			if err != io.EOF {
+				s.err = err
+			}
+			s.done = true
 			s.mu.Unlock()
+			s.signal()
 			return
 		}
-		s.cond.Broadcast()
 		s.mu.Unlock()
+		if n > 0 {
+			s.signal()
+		}
 	}
 }
 
-// ReadKey blocks until a whole key token is available (or EOF), tokenizing one
-// key per call off the shared buffer via scanKey. "" signals end-of-input (the
-// read-key nil contract). Blocking here matches native (which blocks in poll);
-// callers stay responsive by gating on KeyPending first.
+// ReadKey blocks until a whole key token is available (or the reader ends),
+// tokenizing one key per call off the shared buffer via scanKey. It returns ""
+// with a nil error at a clean EOF (the read-key nil contract), or "" with the
+// retained error if the reader failed. Blocking here matches native (which
+// blocks in poll); callers stay responsive by gating on KeyPending first.
 func (s *queuedKeySource) ReadKey() (string, error) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 	s.start()
-	for len(s.buf) == 0 && !s.eof {
-		s.cond.Wait()
+
+	for len(s.buf) == 0 && !s.done {
+		s.waitLocked(0) // no deadline: block until data or reader end
 	}
 	if len(s.buf) == 0 {
-		return "", nil // EOF
+		return "", s.err // clean EOF (err nil) or a real reader failure
 	}
+
+	var deadline time.Time // set on the first incomplete front token
 	for {
 		status, n := scanKey(s.buf)
-		if status == keyNeedMore && n > 1 && !s.eof {
-			// A multi-byte token (CSI/SS3/UTF-8) split across reader chunks —
-			// short reads are normal over a network transport like 9P/drawterm.
-			// Wait for one more append, then re-scan to stitch it, rather than
-			// emitting a broken partial that would desync every following key.
-			// Bounded: each wait adds bytes (progress) or hits eof (then emit
-			// best-effort below). A lone ESC (scanKey → keyNeedMore, 1) is
-			// deliberately NOT waited on — it is the Escape key, emitted now
-			// rather than parked until the next keypress.
-			before := len(s.buf)
-			for len(s.buf) == before && !s.eof {
-				s.cond.Wait()
-			}
-			continue
+		if status == keyReady || s.done {
+			// Complete token, or the reader ended so no completion is coming —
+			// emit best-effort (a lone ESC at EOF is the Escape key; a truncated
+			// sequence surfaces rather than hangs).
+			return s.take(n), nil
 		}
-		tok := string(s.buf[:n])
-		s.buf = s.buf[n:]
-		if len(s.buf) == 0 {
-			s.buf = nil // release the backing array once drained
+		// keyNeedMore and the reader is still live: a front token split across
+		// reads. Wait for more bytes up to the grace period, then re-scan; if the
+		// grace expires with no completion, emit what we have. This is what makes
+		// a bare ESC vs. a split arrow decidable without a FIONREAD peek. The
+		// grace is inter-byte — each byte that arrives resets it — so a sequence
+		// dribbling in over a slow transport keeps stitching, while a truly bare
+		// ESC (nothing more within one grace window) emits as the Escape key.
+		if deadline.IsZero() {
+			deadline = time.Now().Add(s.grace)
 		}
-		return tok, nil
+		remaining := time.Until(deadline)
+		if remaining <= 0 {
+			return s.take(n), nil
+		}
+		before := len(s.buf)
+		s.waitLocked(remaining)
+		if len(s.buf) > before {
+			deadline = time.Time{} // new bytes — restart the inter-byte window
+		}
 	}
+}
+
+// take slices the front n bytes off the buffer as a token, releasing the backing
+// array once drained. Called with s.mu held.
+func (s *queuedKeySource) take(n int) string {
+	tok := string(s.buf[:n])
+	s.buf = s.buf[n:]
+	if len(s.buf) == 0 {
+		s.buf = nil
+	}
+	return tok
+}
+
+// waitLocked releases the lock, waits for a reader wake (or the timeout, if
+// nonzero), and re-acquires it — the condition-variable wait, done over a
+// channel so it can be time-bounded. Called with s.mu held; returns holding it.
+func (s *queuedKeySource) waitLocked(timeout time.Duration) {
+	s.mu.Unlock()
+	if timeout > 0 {
+		select {
+		case <-s.notify:
+		case <-time.After(timeout):
+		}
+	} else {
+		<-s.notify
+	}
+	s.mu.Lock()
 }
 
 // KeyPending reports whether a key is buffered and ready, without consuming it.

--- a/pkg/rt/keysource_queued_test.go
+++ b/pkg/rt/keysource_queued_test.go
@@ -1,0 +1,175 @@
+/*
+ * Copyright (c) 2026 Matt Parrett
+ * SPDX-License-Identifier: MIT
+ */
+
+package rt
+
+import (
+	"io"
+	"reflect"
+	"testing"
+	"time"
+)
+
+// chunkReader hands out one queued chunk per Read (splitting a chunk across
+// Reads only if the caller's buffer is smaller), blocks when the queue is empty,
+// and returns io.EOF once the channel is closed. It lets a test control exactly
+// how input bytes are split across reads — the thing the queued source has to
+// stitch back together.
+type chunkReader struct {
+	ch  chan []byte
+	rem []byte
+}
+
+func (c *chunkReader) Read(p []byte) (int, error) {
+	if len(c.rem) == 0 {
+		b, ok := <-c.ch
+		if !ok {
+			return 0, io.EOF
+		}
+		c.rem = b
+	}
+	n := copy(p, c.rem)
+	c.rem = c.rem[n:]
+	return n, nil
+}
+
+// feed returns a queued source over the given chunks, delivered in order then
+// EOF. All chunks are queued up front, so ReadKey sees end-of-input after the
+// last one.
+func feed(chunks ...[]byte) KeySource {
+	ch := make(chan []byte, len(chunks))
+	for _, c := range chunks {
+		ch <- c
+	}
+	close(ch)
+	return NewQueuedKeySource(&chunkReader{ch: ch})
+}
+
+// readAll drains keys until the EOF nil contract ("") and returns the tokens.
+// Guarded by a deadline so a stuck ReadKey fails loudly instead of hanging CI.
+func readAll(t *testing.T, ks KeySource) []string {
+	t.Helper()
+	var got []string
+	done := make(chan []string, 1)
+	go func() {
+		var out []string
+		for {
+			k, err := ks.ReadKey()
+			if err != nil {
+				t.Errorf("ReadKey error: %v", err)
+				break
+			}
+			if k == "" {
+				break
+			}
+			out = append(out, k)
+		}
+		done <- out
+	}()
+	select {
+	case got = <-done:
+	case <-time.After(2 * time.Second):
+		t.Fatal("readAll timed out — ReadKey blocked")
+	}
+	return got
+}
+
+func TestQueuedKeySourceTokenizes(t *testing.T) {
+	cases := []struct {
+		name   string
+		chunks [][]byte
+		want   []string
+	}{
+		{"held-key burst", [][]byte{[]byte("llll")}, []string{"l", "l", "l", "l"}},
+		{"arrow one chunk", [][]byte{[]byte("\x1b[A")}, []string{"\x1b[A"}},
+		{"lone esc", [][]byte{[]byte("\x1b")}, []string{"\x1b"}},
+		{"ss3 f1", [][]byte{[]byte("\x1bOP")}, []string{"\x1bOP"}},
+		{"mixed keys and csi", [][]byte{[]byte("l\x1b[Bx")}, []string{"l", "\x1b[B", "x"}},
+		// A CSI arrow split across two reads must come back as one token.
+		{"split arrow", [][]byte{[]byte("\x1b["), []byte("A")}, []string{"\x1b[A"}},
+		// A multi-byte rune split across reads must never be torn.
+		{"split utf8", [][]byte{{0xe2}, {0x8c, 0x98}}, []string{"⌘"}},
+		// A truncated CSI at EOF is emitted best-effort rather than lost.
+		{"truncated csi at eof", [][]byte{[]byte("\x1b[")}, []string{"\x1b["}},
+		{"empty then eof", nil, nil},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			got := readAll(t, feed(tc.chunks...))
+			if !reflect.DeepEqual(got, tc.want) {
+				t.Errorf("tokens = %q, want %q", got, tc.want)
+			}
+		})
+	}
+}
+
+// TestQueuedKeySourceStitchesWhileBlocked drives the wait-for-more path
+// deterministically: the partial CSI arrives while ReadKey is already parked, so
+// ReadKey must wait for the completing byte and stitch, not emit "\x1b[".
+func TestQueuedKeySourceStitchesWhileBlocked(t *testing.T) {
+	ch := make(chan []byte) // unbuffered: each send blocks until the reader takes it
+	ks := NewQueuedKeySource(&chunkReader{ch: ch})
+
+	res := make(chan string, 1)
+	go func() {
+		k, _ := ks.ReadKey()
+		res <- k
+	}()
+
+	ch <- []byte("\x1b[") // reader appends; ReadKey wakes, sees keyNeedMore, waits
+	ch <- []byte("A")     // completes the sequence; ReadKey stitches and returns
+
+	select {
+	case got := <-res:
+		if got != "\x1b[A" {
+			t.Errorf("stitched token = %q, want %q", got, "\x1b[A")
+		}
+	case <-time.After(2 * time.Second):
+		t.Fatal("ReadKey did not return after the sequence completed")
+	}
+	close(ch)
+}
+
+// TestQueuedKeySourceKeyPending verifies the non-blocking, eof-blind peek:
+// false before input, true once a byte is buffered, false again once drained at
+// end-of-input (so a poll loop doesn't busy-spin on EOF).
+func TestQueuedKeySourceKeyPending(t *testing.T) {
+	ch := make(chan []byte, 1)
+	ks := NewQueuedKeySource(&chunkReader{ch: ch})
+
+	// Starts the reader; nothing queued yet → not pending.
+	if ks.KeyPending() {
+		t.Fatal("KeyPending true before any input")
+	}
+
+	ch <- []byte("x")
+	if !eventually(func() bool { return ks.KeyPending() }) {
+		t.Fatal("KeyPending never became true after input")
+	}
+
+	if k, _ := ks.ReadKey(); k != "x" {
+		t.Fatalf("ReadKey = %q, want %q", k, "x")
+	}
+	if ks.KeyPending() {
+		t.Fatal("KeyPending true after the only key was consumed")
+	}
+
+	// EOF must not read as pending (else xsofy's input poll busy-drains it).
+	close(ch)
+	if eventually(func() bool { return ks.KeyPending() }) {
+		t.Fatal("KeyPending true at EOF")
+	}
+}
+
+func eventually(cond func() bool) bool {
+	deadline := time.Now().Add(time.Second)
+	for time.Now().Before(deadline) {
+		if cond() {
+			return true
+		}
+		time.Sleep(time.Millisecond)
+	}
+	return cond()
+}

--- a/pkg/rt/keysource_queued_test.go
+++ b/pkg/rt/keysource_queued_test.go
@@ -6,6 +6,7 @@
 package rt
 
 import (
+	"errors"
 	"io"
 	"reflect"
 	"testing"
@@ -35,6 +36,15 @@ func (c *chunkReader) Read(p []byte) (int, error) {
 	return n, nil
 }
 
+// testGrace is a short inter-byte grace so timing-dependent cases resolve fast.
+const testGrace = 15 * time.Millisecond
+
+// newSource builds a queued source over r with the test grace. Constructs the
+// unexported type directly (same package) so tests can shorten the grace.
+func newSource(r io.Reader) *queuedKeySource {
+	return &queuedKeySource{r: r, notify: make(chan struct{}, 1), grace: testGrace}
+}
+
 // feed returns a queued source over the given chunks, delivered in order then
 // EOF. All chunks are queued up front, so ReadKey sees end-of-input after the
 // last one.
@@ -44,7 +54,7 @@ func feed(chunks ...[]byte) KeySource {
 		ch <- c
 	}
 	close(ch)
-	return NewQueuedKeySource(&chunkReader{ch: ch})
+	return newSource(&chunkReader{ch: ch})
 }
 
 // readAll drains keys until the EOF nil contract ("") and returns the tokens.
@@ -105,31 +115,101 @@ func TestQueuedKeySourceTokenizes(t *testing.T) {
 	}
 }
 
-// TestQueuedKeySourceStitchesWhileBlocked drives the wait-for-more path
-// deterministically: the partial CSI arrives while ReadKey is already parked, so
-// ReadKey must wait for the completing byte and stitch, not emit "\x1b[".
-func TestQueuedKeySourceStitchesWhileBlocked(t *testing.T) {
-	ch := make(chan []byte) // unbuffered: each send blocks until the reader takes it
-	ks := NewQueuedKeySource(&chunkReader{ch: ch})
+// TestQueuedKeySourceStitchesAcrossReads drives the wait-for-more path
+// deterministically: an arrow arrives byte-by-byte over an open channel — the
+// worst case the review called out (ESC, then [, then A as separate reads).
+// ReadKey must stitch it into one token, not emit ESC early and split the rest.
+func TestQueuedKeySourceStitchesAcrossReads(t *testing.T) {
+	for _, tc := range []struct {
+		name   string
+		chunks []string
+		want   string
+	}{
+		{"esc bracket A separately", []string{"\x1b", "[", "A"}, "\x1b[A"},
+		{"esc then bracket-A", []string{"\x1b", "[A"}, "\x1b[A"},
+		{"esc-bracket then A", []string{"\x1b[", "A"}, "\x1b[A"},
+		{"utf8 byte by byte", []string{"\xe2", "\x8c", "\x98"}, "⌘"},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			ch := make(chan []byte) // unbuffered: each send blocks until read
+			ks := newSource(&chunkReader{ch: ch})
+			res := make(chan string, 1)
+			go func() {
+				k, _ := ks.ReadKey()
+				res <- k
+			}()
+			for _, c := range tc.chunks {
+				ch <- []byte(c) // each byte arrives within the inter-byte grace
+			}
+			select {
+			case got := <-res:
+				if got != tc.want {
+					t.Errorf("stitched token = %q, want %q", got, tc.want)
+				}
+			case <-time.After(2 * time.Second):
+				t.Fatal("ReadKey did not return after the sequence completed")
+			}
+			close(ch)
+		})
+	}
+}
+
+// TestQueuedKeySourceBareEscEmitsAfterGrace: a lone ESC with no completing byte
+// (channel stays open, so no EOF short-circuit) must resolve to the Escape key
+// once the inter-byte grace expires — not hang, and not split.
+func TestQueuedKeySourceBareEscEmitsAfterGrace(t *testing.T) {
+	ch := make(chan []byte, 1)
+	ks := newSource(&chunkReader{ch: ch})
+	ch <- []byte("\x1b") // no close — reader blocks after this, done stays false
 
 	res := make(chan string, 1)
 	go func() {
 		k, _ := ks.ReadKey()
 		res <- k
 	}()
-
-	ch <- []byte("\x1b[") // reader appends; ReadKey wakes, sees keyNeedMore, waits
-	ch <- []byte("A")     // completes the sequence; ReadKey stitches and returns
-
 	select {
 	case got := <-res:
-		if got != "\x1b[A" {
-			t.Errorf("stitched token = %q, want %q", got, "\x1b[A")
+		if got != "\x1b" {
+			t.Errorf("bare ESC = %q, want %q", got, "\x1b")
 		}
 	case <-time.After(2 * time.Second):
-		t.Fatal("ReadKey did not return after the sequence completed")
+		t.Fatal("bare ESC never resolved — likely parked waiting for more bytes")
 	}
 	close(ch)
+}
+
+// TestQueuedKeySourceReaderErrorPropagates: a non-EOF reader error must surface
+// through ReadKey's error return after buffered bytes drain, not be swallowed as
+// a clean EOF.
+func TestQueuedKeySourceReaderErrorPropagates(t *testing.T) {
+	wantErr := errors.New("console read failed")
+	ks := newSource(&errAfterReader{data: []byte("ab"), err: wantErr})
+
+	// The bytes that arrived before the error still tokenize.
+	for _, want := range []string{"a", "b"} {
+		if k, err := ks.ReadKey(); k != want || err != nil {
+			t.Fatalf("ReadKey = (%q, %v), want (%q, nil)", k, err, want)
+		}
+	}
+	// Then the retained error surfaces.
+	if k, err := ks.ReadKey(); k != "" || err != wantErr {
+		t.Fatalf("ReadKey at error = (%q, %v), want (\"\", %v)", k, err, wantErr)
+	}
+}
+
+// errAfterReader yields data once, then returns err on the next read.
+type errAfterReader struct {
+	data []byte
+	err  error
+	done bool
+}
+
+func (r *errAfterReader) Read(p []byte) (int, error) {
+	if !r.done {
+		r.done = true
+		return copy(p, r.data), nil
+	}
+	return 0, r.err
 }
 
 // TestQueuedKeySourceKeyPending verifies the non-blocking, eof-blind peek:

--- a/pkg/rt/term_plan9.go
+++ b/pkg/rt/term_plan9.go
@@ -10,15 +10,40 @@ package rt
 import (
 	"fmt"
 	"os"
+	"strconv"
+	"sync"
 	"unicode/utf8"
 
 	"github.com/nooga/let-go/pkg/vm"
 )
 
-// Plan 9 has no termios/ioctl. raw-mode/size/pty are stubbed; the ANSI output
-// functions route through *out* (WriteToOut) like native (rio won't render
-// escapes, but they're harmless). If you need real terminal control on plan9,
-// wire in /dev/cons here.
+// Plan 9 terminal support. Input is real: raw-mode! holds /dev/consctl open with
+// "rawon" (Plan 9 has no termios; consctl controls the console), and read-key /
+// key-pending? read through a queuedKeySource (keysource_queued.go) — a
+// background goroutine feeding a byte queue, because Plan 9 lacks the Unix
+// poll(2)/FIONREAD non-blocking peek. The ANSI output functions route through
+// *out* (WriteToOut) like native. size reads $COLS/$LINES (fallback 80x24); real
+// window geometry (/dev/wctl pixels) is a deferred follow-up. open-pty/set-size
+// stay unsupported. mouse decoding is native-only.
+
+// consctl holds /dev/consctl open while the terminal is in raw mode. Plan 9
+// reverts to cooked mode as soon as this fd closes, so raw-mode! keeps it open
+// and restore-mode! writes "rawoff" then closes it. Guarded by consctlMu.
+var (
+	consctl   *os.File
+	consctlMu sync.Mutex
+)
+
+// envInt reads an integer environment variable, returning def when it is unset
+// or unparseable.
+func envInt(name string, def int) int {
+	if v := os.Getenv(name); v != "" {
+		if n, err := strconv.Atoi(v); err == nil {
+			return n
+		}
+	}
+	return def
+}
 
 func init() { RegisterInstaller(installTermNS) }
 
@@ -32,23 +57,97 @@ func installTermNS() {
 	ns := vm.NewNamespace("term")
 	ns.Refer(CoreNS, "", true)
 
-	stubTrue, _ := vm.NativeFnType.Wrap(func(vs []vm.Value) (vm.Value, error) {
-		return vm.TRUE, nil
-	})
-	stubFalse, _ := vm.NativeFnType.Wrap(func(vs []vm.Value) (vm.Value, error) {
-		return vm.FALSE, nil
-	})
 	stubNil, _ := vm.NativeFnType.Wrap(func(vs []vm.Value) (vm.Value, error) {
 		return vm.NIL, nil
 	})
 
-	ns.Def("raw-mode!", stubTrue)
-	ns.Def("restore-mode!", stubTrue)
-	ns.Def("read-key", stubNil)
-	ns.Def("key-pending?", stubFalse)
+	// Bind the plan9 key source (background stdin reader + queue) at the *keys*
+	// root — the input dual of the os.Stdout *out* default in iort.go. read-key
+	// and key-pending? consult the bound source, so api.WithKeySource / (binding
+	// [*keys* …]) transparently overrides it for tests and embedders.
+	CoreNS.Lookup("*keys*").(*vm.Var).SetRoot(vm.NewBoxed(NewQueuedKeySource(os.Stdin)))
 
+	// raw-mode! — enter raw mode by opening /dev/consctl and writing "rawon",
+	// holding the fd (Plan 9 reverts to cooked mode when it closes). Only the
+	// 0-arg global form is supported; xsofy never passes a handle. Returns nil
+	// (not an error) when there's no console — mirrors native's not-a-TTY path.
+	rawMode, _ := vm.NativeFnType.Wrap(func(vs []vm.Value) (vm.Value, error) {
+		if len(vs) != 0 {
+			return vm.NIL, fmt.Errorf("raw-mode!: per-handle raw mode not supported on plan9")
+		}
+		consctlMu.Lock()
+		defer consctlMu.Unlock()
+		if consctl != nil {
+			return vm.TRUE, nil // already raw
+		}
+		f, err := os.OpenFile("/dev/consctl", os.O_WRONLY, 0)
+		if err != nil {
+			return vm.NIL, nil // no console (e.g. piped stdin)
+		}
+		if _, err := f.WriteString("rawon"); err != nil {
+			f.Close()
+			return vm.NIL, nil
+		}
+		consctl = f
+		return vm.TRUE, nil
+	})
+	ns.Def("raw-mode!", rawMode)
+
+	// restore-mode! — write "rawoff" and close /dev/consctl, reverting to cooked
+	// mode. Idempotent: a no-op (nil) when raw mode was never entered.
+	restoreMode, _ := vm.NativeFnType.Wrap(func(vs []vm.Value) (vm.Value, error) {
+		consctlMu.Lock()
+		defer consctlMu.Unlock()
+		if consctl == nil {
+			return vm.NIL, nil
+		}
+		_, _ = consctl.WriteString("rawoff")
+		err := consctl.Close()
+		consctl = nil
+		if err != nil {
+			return vm.NIL, fmt.Errorf("restore-mode!: %w", err)
+		}
+		return vm.TRUE, nil
+	})
+	ns.Def("restore-mode!", restoreMode)
+
+	// read-key — read one keypress through the *keys* source. Returns single
+	// chars, or escape sequences like "\x1b[A" for arrows; "" is the
+	// end-of-input nil contract. Ctx-aware so api.WithKeySource is honored.
+	// Mouse-report decoding is native-only — plan9 has no mouse path.
+	readKey := vm.NewCtxNativeFn("read-key", func(ec *vm.ExecContext, vs []vm.Value) (vm.Value, error) {
+		s, err := boundKeySource(ec).ReadKey()
+		if err != nil {
+			return vm.NIL, err
+		}
+		if s == "" {
+			return vm.NIL, nil
+		}
+		return vm.String(s), nil
+	})
+	ns.Def("read-key", readKey)
+
+	// key-pending? — true if a key is buffered and ready, without consuming it.
+	// Plan 9 has no poll/FIONREAD, so the bound source answers from its
+	// background-reader queue. Non-blocking; eof-blind so the
+	// (when (key-pending?) (read-key)) idiom doesn't busy-spin at end-of-input.
+	keyPendingFn := vm.NewCtxNativeFn("key-pending?", func(ec *vm.ExecContext, vs []vm.Value) (vm.Value, error) {
+		if boundKeySource(ec).KeyPending() {
+			return vm.TRUE, nil
+		}
+		return vm.FALSE, nil
+	})
+	ns.Def("key-pending?", keyPendingFn)
+
+	// size — [cols rows] from $COLS/$LINES, falling back to 80x24. Real Plan 9
+	// window geometry lives in /dev/wctl (pixels) and is rarely surfaced as env
+	// vars under rio/drawterm, so this usually returns the fallback — safe,
+	// since callers diff term/size per tick and simply see "no resize".
 	sizeFn, _ := vm.NativeFnType.Wrap(func(vs []vm.Value) (vm.Value, error) {
-		return vm.NewPersistentVector([]vm.Value{vm.MakeInt(80), vm.MakeInt(24)}), nil
+		return vm.NewPersistentVector([]vm.Value{
+			vm.MakeInt(envInt("COLS", 80)),
+			vm.MakeInt(envInt("LINES", 24)),
+		}), nil
 	})
 	ns.Def("size", sizeFn)
 

